art/estimators/object_detection/pytorch_yolo.py

```diff
@@ -256,149 +267,186 @@ def device(self) -> "torch.device":
         """
         return self._device
 
-    def _get_losses(
-        self, x: Union[np.ndarray, "torch.Tensor"], y: List[Dict[str, Union[np.ndarray, "torch.Tensor"]]]
-    ) -> Tuple[Dict[str, "torch.Tensor"], "torch.Tensor", "torch.Tensor"]:
+    def _preprocess_and_convert_inputs(
+        self,
+        x: Union[np.ndarray, "torch.Tensor"],
+        y: Optional[List[Dict[str, Union[np.ndarray, "torch.Tensor"]]]] = None,
+        fit: bool = False,
+        no_grad: bool = True,
+    ) -> Tuple["torch.Tensor", List[Dict[str, "torch.Tensor"]]]:
         """
-        Get the loss tensor output of the model including all preprocessing.
+        Apply preprocessing on inputs `(x, y)` and convert to tensors, if needed.
 
-        :param x: Samples of shape (nb_samples, height, width, nb_channels).
-        :param y: Target values of format `List[Dict[Tensor]]`, one for each input image. The fields of the Dict are as
-                  follows:
-
-                  - boxes (FloatTensor[N, 4]): the boxes in [x1, y1, x2, y2] format, with 0 <= x1 < x2 <= W and
-                                               0 <= y1 < y2 <= H.
-                  - labels (Int64Tensor[N]): the labels for each image
-        :return: Loss gradients of the same shape as `x`.
+        :param x: Samples of shape NCHW or NHWC.
+        :param y: Target values of format `List[Dict[str, Union[np.ndarray, torch.Tensor]]]`, one for each input image.
+                  The fields of the Dict are as follows:
+
+                  - boxes [N, 4]: the boxes in [x1, y1, x2, y2] format, with 0 <= x1 < x2 <= W and 0 <= y1 < y2 <= H.
+                  - labels [N]: the labels for each image.
+        :param fit: `True` if the function is call before fit/training and `False` if the function is called before a
+                    predict operation.
+        :param no_grad: `True` if no gradients required.
+        :return: Preprocessed inputs `(x, y)` as tensors.
         """
         import torch
 
-        self._model.train()
-        self.set_batchnorm(train=False)
-        self.set_dropout(train=False)
```

Comment:
Where are  `set_batchnorm` and `set_dropout` applied now?

I think we might have set them in `loss_gradient`, `predict` and `fit` to make sure these layers are set correctly.

Reply:
Based on our discussion, the `set_batchnorm` and `set_dropout` should no longer be needed as subsequent calls to the `loss_gradient` yield gradients that are still the same within numerical accuracy.